From: local
Subject: [PATCH] dwmblocks statuscmd matched to dwm SIGRTMIN protocol

Prefix each block with its signal byte. On SIGRTMIN+N with a button
value, rerun that block with BLOCK_BUTTON and BUTTON set. A bare
pkill -RTMIN+N still refreshes.

--- a/dwmblocks.c
+++ b/dwmblocks.c
@@ -25,12 +25,13 @@
 	unsigned int signal;
 } Block;
 #ifndef __OpenBSD__
 void dummysighandler(int num);
 #endif
 void sighandler(int num);
+void buttonhandler(int sig, siginfo_t *si, void *ucontext);
 void getcmds(int time);
 void getsigcmds(unsigned int signal);
 void setupsignals();
 void sighandler(int signum);
 int getstatus(char *str, char *last);
 void statusloop();
@@ -49,26 +50,41 @@
 
 
 #include "blocks.h"
 
 static char statusbar[LENGTH(blocks)][CMDLENGTH] = {0};
 static char statusstr[2][STATUSLENGTH];
+static char button[] = "\0";
 static int statusContinue = 1;
 //static int returnStatus = 0;
 
 //opens process *cmd and stores output in *output
 void getcmd(const Block *block, char *output)
 {
 	//make sure status is same until output is ready
 	char tempstatus[CMDLENGTH] = {0};
-	strcpy(tempstatus, block->icon);
-	FILE *cmdf = popen(block->command, "r");
+	int off = 0;
+	FILE *cmdf;
+
+	if (block->signal)
+		tempstatus[off++] = block->signal;
+	strcpy(tempstatus + off, block->icon);
+	if (*button) {
+		setenv("BLOCK_BUTTON", button, 1);
+		setenv("BUTTON", button, 1);
+		cmdf = popen(block->command, "r");
+		*button = '\0';
+		unsetenv("BLOCK_BUTTON");
+		unsetenv("BUTTON");
+	} else {
+		cmdf = popen(block->command, "r");
+	}
 	if (!cmdf)
 		return;
 	int i = strlen(block->icon);
-	fgets(tempstatus+i, CMDLENGTH-i-delimLen, cmdf);
+	fgets(tempstatus + off + i, CMDLENGTH - off - i - delimLen, cmdf);
 	i = strlen(tempstatus);
 	//if block and command output are both not empty
 	if (i != 0) {
 		//only chop off newline if one is present at the end
 		i = tempstatus[i-1] == '\n' ? i-1 : i;
 		if (delim[0] != '\0') {
@@ -100,21 +116,29 @@
 			getcmd(current,statusbar[i]);
 	}
 }
 
 void setupsignals()
 {
+	struct sigaction sa;
+
+	sigemptyset(&sa.sa_mask);
+	sa.sa_sigaction = buttonhandler;
+	sa.sa_flags = SA_SIGINFO;
 #ifndef __OpenBSD__
-	    /* initialize all real time signals with dummy handler */
-    for (int i = SIGRTMIN; i <= SIGRTMAX; i++)
-        signal(i, dummysighandler);
+	/* initialize all real time signals with dummy handler */
+	for (int i = SIGRTMIN; i <= SIGRTMAX; i++)
+		signal(i, dummysighandler);
 #endif
 
-	for (unsigned int i = 0; i < LENGTH(blocks); i++)
-		if (blocks[i].signal > 0)
-			signal(SIGMINUS+blocks[i].signal, sighandler);
+	for (unsigned int i = 0; i < LENGTH(blocks); i++) {
+		if (blocks[i].signal > 0) {
+			sigaddset(&sa.sa_mask, SIGRTMIN + blocks[i].signal);
+			sigaction(SIGRTMIN + blocks[i].signal, &sa, NULL);
+		}
+	}
 }
 
 int getstatus(char *str, char *last)
 {
 	strcpy(last, str);
 	str[0] = '\0';
@@ -172,12 +196,22 @@
 #ifndef __OpenBSD__
 /* this signal handler should do nothing */
 void dummysighandler(int signum)
 {
     return;
 }
+
+void buttonhandler(int sig, siginfo_t *si, void *ucontext)
+{
+	int btn = si ? si->si_value.sival_int : 0;
+
+	if (btn)
+		*button = '0' + (btn & 0xff);
+	getsigcmds(sig - SIGRTMIN);
+	writestatus();
+}
 #endif
 
 void sighandler(int signum)
 {
 	getsigcmds(signum-SIGPLUS);
 	writestatus();
